File: cpp/example_code/ec2/ec2_describe_regions_and_zones.cpp
/*
   Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.

   This file is licensed under the Apache License, Version 2.0 (the "License").
   You may not use this file except in compliance with the License. A copy of
   the License is located at

    http://aws.amazon.com/apache2.0/

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied. See the License for the
   specific language governing permissions and limitations under the License.
*/
#include <aws/core/Aws.h>

#include <aws/ec2/EC2Client.h>
#include <aws/ec2/model/DescribeAvailabilityZonesRequest.h>
#include <aws/ec2/model/DescribeAvailabilityZonesResponse.h>
#include <aws/ec2/model/DescribeRegionsRequest.h>
#include <aws/ec2/model/DescribeRegionsResponse.h>

#include <iostream>

/**
 * Describes all regions and zones
 */
int main(int argc, char** argv)
{
    Aws::SDKOptions options;
    Aws::InitAPI(options);

    Aws::EC2::EC2Client ec2_client;

    Aws::EC2::Model::DescribeRegionsRequest describeRegionsRequest;

    auto describeRegionsOutcome = ec2_client.DescribeRegions(describeRegionsRequest);
    if(describeRegionsOutcome.IsSuccess())
    {
        std::cout << std::left << std::setw(32) << "RegionName" 
                               << std::setw(64) << "Endpoint" << std::endl;

        const auto& regions = describeRegionsOutcome.GetResult().GetRegions();
        for(const auto& region : regions)
        {
            std::cout << std::left << std::setw(32) << region.GetRegionName() 
                                   << std::setw(64) << region.GetEndpoint() << std::endl;
        }
    }
    else
    {
        std::cout << "Failed to describe regions:" << describeRegionsOutcome.GetError().GetMessage() << std::endl;
    }

    std::cout << std::endl;

    Aws::EC2::Model::DescribeAvailabilityZonesRequest describeAvailabilityZonesRequest;

    auto describeAvailabilityZonesOutcome = ec2_client.DescribeAvailabilityZones(describeAvailabilityZonesRequest);
    if(describeAvailabilityZonesOutcome.IsSuccess())
    {
        std::cout << std::left << std::setw(32) << "ZoneName"
                               << std::setw(20) << "State"
                               << std::setw(32) << "Region" << std::endl;

        const auto& availabilityZones = describeAvailabilityZonesOutcome.GetResult().GetAvailabilityZones();
        for(const auto& availabilityZone : availabilityZones)
        {
            Aws::String stateString = Aws::EC2::Model::AvailabilityZoneStateMapper::GetNameForAvailabilityZoneState(availabilityZone.GetState());
            std::cout << std::left << std::setw(32) << availabilityZone.GetZoneName()
                                   << std::setw(20) << stateString 
                                   << std::setw(32) << availabilityZone.GetRegionName() << std::endl;
        }
    }
    else
    {
        std::cout << "Failed to describe availability zones:" << describeAvailabilityZonesOutcome.GetError().GetMessage() << std::endl;
    }

    Aws::ShutdownAPI(options);

    return 0;
}

File: javascript/example_code/ec2/ec2_describeregionsandzones.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

// ABOUT THIS NODE.JS SAMPLE: This sample is part of the SDK for JavaScript Developer Guide topic at
// https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/ec2-example-security-groups.html

// snippet-start:[ec2.JavaScript.Regions.describeRegions]
// Load the AWS SDK for Node.js
var AWS = require("aws-sdk");
// Set the region
AWS.config.update({ region: "REGION" });

// Create EC2 service object
var ec2 = new AWS.EC2({ apiVersion: "2016-11-15" });

var params = {};

// Retrieves all regions/endpoints that work with EC2
ec2.describeRegions(params, function (err, data) {
  if (err) {
    console.log("Error", err);
  } else {
    console.log("Regions: ", data.Regions);
  }
});

// Retrieves availability zones only for region of the ec2 service object
ec2.describeAvailabilityZones(params, function (err, data) {
  if (err) {
    console.log("Error", err);
  } else {
    console.log("Availability Zones: ", data.AvailabilityZones);
  }
});
// snippet-end:[ec2.JavaScript.Regions.describeRegions]
